lib/active_record/pg_enum/postgresql_adapter.rb

```diff
@@ -13,29 +13,46 @@ module PostgreSQLAdapter
       #
       #   { "foo_type" => ["foo", "bar", "baz"] }
       def enum_types
-        res = exec_query(<<-SQL.strip_heredoc, "SCHEMA")
-          SELECT t.typname AS enum_name, array_agg(e.enumlabel ORDER BY e.enumsortorder) AS enum_value
-          FROM pg_type t
-          JOIN pg_enum e ON t.oid = e.enumtypid
-          JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
-          WHERE n.nspname = ANY (current_schemas(false))
-          GROUP BY enum_name;
-        SQL
+        pg_type = Arel::Table.new(:pg_type)
+        pg_enum = Arel::Table.new(:pg_enum)
+        pg_namespace = Arel::Table.new(:pg_namespace)
+        current_schema = Arel::Nodes::NamedFunction.new('current_schemas', [Arel::Nodes::False.new])
+        any_current_schema = Arel::Nodes::NamedFunction.new('ANY', [current_schema])
 
-        coltype = res.column_types["enum_value"]
-        deserialize = if coltype.respond_to?(:deserialize)
-          proc { |values| coltype.deserialize(values) }
-        elsif coltype.respond_to?(:type_cast_from_database)
-          proc { |values| coltype.type_cast_from_database(values) }
-        else
-          proc { |values| coltype.type_cast(values) }
+        query = pg_type.
+          project(pg_type[:typname], pg_enum[:enumlabel]).
+          join(pg_enum).on(pg_enum[:enumtypid].eq(pg_type[:oid])).
+          join(pg_namespace).on(pg_type[:typnamespace].eq(pg_namespace[:oid])).
+          where(pg_namespace[:nspname].eq(any_current_schema)).
+          order(pg_type[:typname], pg_enum[:enumsortorder])
+        rows = ActiveRecord::Base.connection.select_rows(query.to_sql)
+
+        rows.reduce({}) do |accum, (type, value)|
+          accum[type.to_s] ||= []
+          accum[type.to_s] << value.to_s
+          accum
         end
+      end
+
+      # Helper method that returns only the values of a specific ENUM types.  Uses a WeakRef as a short lived
+      # cache to prevent hitting the database for every enum on startup.
+      #
+      # Returns an array of strings like ["foo", "bar", "baz"]
+      def enum_values(type_name)
+        @_cached_enum_types ||= WeakRef.new(enum_types)
+        @_cached_enum_types[type_name.to_s]
+      end
 
-        res.rows
-          .map { |name, values| [name, values] }
-          .sort { |a, b| a.first <=> b.first }
-          .each_with_object({}) { |(name, values), memo| memo[name] = deserialize.call(values) }
+      # Helper method to creates an ActiveRecord enum, inferring the values from the ENUM type specified.
+      #
+      # Can be called as `postgres_enum(:foo, :foo_type)`
+      #
+      def postgres_enum(name, type_name = nil)
+        type_name ||= name
+        values = enum_values(type_name)
+        enum name.to_sym => values.zip(values)
       end
```

Comment:
I've made an intentional design choice to avoid adding to the public API of anyone's models, which is why I wrote the `PGEnum()` module builder.

I _might_ be convinced to add a model concern for automatic wireup, but as I said in the README... I would prefer that you can see what the enum values are by looking at the model.

Reply:
So two things.  

1.  The include is extremely awkward, converting it to be a method is much cleaner.
2.  There are many use cases where there are a lot of enum values, which make using declaring it in both the db and the model painful.  

I already have a change to this PR that accepts either an array or a symbol and I'll resubmit.

Reply:
I think including a module is a perfectly natural way to add behavior to a class, moreso than relying on singleton methods to magically appear. It's also important to me to give users complete control over when and where my API helpers appear in their code.

That said, I'm willing to approve an optional module for this helper that can be included in `ApplicationRecord`. So if you'd like to extract this behavior into `PGEnum::Helper` in a subsequent PR I'll be more receptive.